rust/xaynet-server/src/state_machine/phases/sum.rs

```diff
@@ -40,18 +45,42 @@ where
     C: CoordinatorStorage,
     M: ModelStorage,
 {
-    /// Handles a [`StateMachineRequest`].
+    /// Handles a sum request.
     ///
-    /// If the request is a [`StateMachineRequest::Update`] or
-    /// [`StateMachineRequest::Sum2`] request, the request sender will receive a
-    /// [`RequestError::MessageRejected`].
+    /// # Errors
+    /// Fails if the sum participant cannot be added due to a PET or [`StorageError`].
     async fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), RequestError> {
-        match req {
-            StateMachineRequest::Sum(sum_req) => self.handle_sum(sum_req).await.map(|res| {
-                self.increment_message_metric();
-                res
-            }),
-            _ => Err(RequestError::MessageRejected),
+        // discard if `max_sum_count` is reached
+        if self.has_overmuch_sums() {
+            self.private.discarded += 1;
+            self.increment_message_metric(Measurement::MessageDiscarded);
+            return Ok(());
+        }
+
+        // reject if not a `sum` message
+        let SumRequest {
+            participant_pk,
+            ephm_pk,
+        } = if let StateMachineRequest::Sum(req) = req {
+            req
+        } else {
+            self.private.rejected += 1;
+            self.increment_message_metric(Measurement::MessageRejected);
+            return Err(RequestError::MessageRejected);
+        };
+
+        // accept if processed successfully, otherwise reject
+        match self.update_sum_dict(participant_pk, ephm_pk).await {
+            ok @ Ok(_) => {
+                self.private.accepted += 1;
+                self.increment_message_metric(Measurement::MessageSum);
+                ok
+            }
+            error @ Err(_) => {
+                self.private.rejected += 1;
+                self.increment_message_metric(Measurement::MessageRejected);
+                error
+            }
```

Comment:
I think we can move most of the code into `process_single`.

```rust
async fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), RequestError> {
        // reject if not a `sum` message
        let SumRequest {
            participant_pk,
            ephm_pk,
        } = if let StateMachineRequest::Sum(req) = req {
            req
        } else {
            return Err(RequestError::MessageRejected);
        };

        // accept if processed successfully, otherwise reject
        let res = self.update_sum_dict(participant_pk, ephm_pk).await;
        if res.is_ok() {
            self.private.accepted += 1;
            self.increment_message_metric();
        }
        res
}
```
```rust
#[async_trait]
pub trait Handler {
    /// Handles a request.
    async fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), RequestError>;
    async fn has_enough_messages(&self) -> bool;
    async fn increment_discarded(&mut self);
    async fn increment_rejected(&mut self);
}
```
```rust
#[async_trait]
impl<C, M> Handler for PhaseState<Sum, C, M>
where
    C: CoordinatorStorage,
    M: ModelStorage,
{
    async fn handle_request(&mut self, req: StateMachineRequest) -> Result<(), RequestError> {
    // ...
    }

    fn has_enough_messages(&self) -> bool {
        self.private.accepted >= self.shared.state.max_sum_count
    }
    
    fn increment_discarded(&mut self) {
         self.private.discarded += 1;
    }

    fn increment_rejected(&mut self) -> bool {
        self.private.rejected += 1;
    }
}
```
```rust
async fn process_single(
        &mut self,
        req: StateMachineRequest,
        span: Span,
        resp_tx: ResponseSender,
    ) {
        let _span_guard = span.enter();
        let res = if self.has_enough_messages() {
            Err(RequestError::TooMany)
        } else {
           self.handle_request(req).await
        }

        if let Err(ref err) = res {
            warn!("failed to handle message: {}", err);
            if matches!(RequestError:: TooMany, _err) {
                self.increment_discarded();
                metric!(
                    Measurement::MessageDiscarded,
                    1,
                    ("round_id", self.shared.state.round_id),
                    ("phase", Self::NAME as u8)
                );
            } else {
                self.increment_rejected();
                metric!(
                    Measurement::MessageRejected,
                    1,
                    ("round_id", self.shared.state.round_id),
                    ("phase", Self::NAME as u8)
                );
            }
        }

        // This may error out if the receiver has already been dropped but
        // it doesn't matter for us.
        let _ = resp_tx.send(res);
    }
```
haven't tested the code though, it may not work that way

Reply:
that's a good idea, i'll give that a try. even though it scatters the metrics calls, it reduced code duplication and codifies things in a common interface, which outweighs the former issue.

Reply:
i agree that we have to change the grafana stuff first, i left a todo in the code at the appropriate place. for now we can just keep the additional methods on the trait and try to simplify it later on.

i also took the opportunity to remove more code duplication wrt to `process_until_enough()`, this is now similar to `process_during()`. as a nice side effect of unifying that we now also have more debug level logging.

regarding your last point, that probably makes sense. we should keep that in mind when we work on the parameter adjustment stategy.

overall it would be good if you have another look at my latest changes, since they are not minor.

Reply:
on it👀